packages/mui-system/src/cssVars/createCssVarsProvider.js

```diff
@@ -62,13 +62,19 @@ export default function createCssVarsProvider(options) {
     const ctx = React.useContext(ColorSchemeContext);
     const nested = !!ctx && !disableNestedContext;
 
-    const scopedTheme = themeProp[themeId];
+    const initialTheme = React.useMemo(() => {
+      if (themeProp) {
+        return themeProp;
+      }
+      return typeof defaultTheme === 'function' ? defaultTheme() : defaultTheme;
+    }, [themeProp]);
+    const scopedTheme = initialTheme[themeId];
```

Comment:
This change will improve the performance when user provides a custom theme because the `defaultTheme` won't be created.